ucp/core.py

```diff
@@ -473,28 +508,25 @@ class Endpoint:
     """
 
     def __init__(
-        self,
-        endpoint,
-        ctx,
-        msg_tag_send,
-        msg_tag_recv,
-        ctrl_tag_send,
-        ctrl_tag_recv,
-        guarantee_msg_order,
+        self, endpoint, ctx, guarantee_msg_order,
     ):
         self._ep = endpoint
         self._ctx = ctx
-        self._msg_tag_send = msg_tag_send
-        self._msg_tag_recv = msg_tag_recv
-        self._ctrl_tag_send = ctrl_tag_send
-        self._ctrl_tag_recv = ctrl_tag_recv
+        self._use_tags = False
         self._guarantee_msg_order = guarantee_msg_order
         self._send_count = 0  # Number of calls to self.send()
         self._recv_count = 0  # Number of calls to self.recv()
         self._finished_recv_count = 0  # Number of returned (finished) self.recv() calls
         self._shutting_down_peer = False  # Told peer to shutdown
         self._close_after_n_recv = None
 
+    def wireup_tags(self, msg_tag_send, msg_tag_recv, ctrl_tag_send, ctrl_tag_recv):
+        self._use_tags = True
+        self._msg_tag_send = msg_tag_send
+        self._msg_tag_recv = msg_tag_recv
+        self._ctrl_tag_send = ctrl_tag_send
+        self._ctrl_tag_recv = ctrl_tag_recv
+
```

Comment:
I suggest that we unify all the tags attributes in a `dict` given as argument to `__init__()` and can be `None` when tags isn't used.

Reply:
Any thoughts about adding tags after the fact? I do that on my test for convince, though I'm not sure I would use it in practice

Reply:
I don't think we should support _tags after the fact_ unless there is a real use case.